### PR TITLE
fix race for global vars

### PR DESCRIPTION
**Pull request by shotasilagadze**

### CODE REVIEW

File: cl/phase1/network/services/batch_signature_verification.go [hunk 1]
```diff
@@ -17,7 +18,7 @@ const (
 )
 
 var (
-	batchCheckInterval          = 500 * time.Millisecond
+	batchCheckInterval          = int64(500)
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/batch_signature_verification.go:21]: can use `atomic.Int64` or `atomic.Pointer[time.Duration]`

File: cl/phase1/network/services/batch_signature_verification.go [hunk 2]
```diff
@@ -40,6 +41,10 @@ type AggregateVerificationData struct {
 	GossipData *sentinel.GossipData
 }
 
+func init() {
+	batchCheckInterval.Store(500)
```
Comment (Giulio2002) [inline cl/phase1/network/services/batch_signature_verification.go:45]: ????

File: cl/phase1/network/services/batch_signature_verification.go [hunk 3]
```diff
@@ -77,7 +82,7 @@ func (b *BatchSignatureVerifier) Start() {
 // When receiving AggregateVerificationData, we simply collect all the signature verification data
 // and verify them together - running all the final functions afterwards
 func (b *BatchSignatureVerifier) start(incoming chan *AggregateVerificationData) {
-	ticker := time.NewTicker(batchCheckInterval)
+	ticker := time.NewTicker(time.Duration(batchCheckInterval.Load()) * time.Millisecond)
```
Comment (Giulio2002) [inline cl/phase1/network/services/batch_signature_verification.go:85]: this is likely wrong why are you atomicizing a constant?

File: cl/sentinel/gossip.go [hunk 4]
```diff
@@ -192,7 +192,7 @@ func (s *Sentinel) forkWatcher() {
 		log.Error("[Gossip] Failed to calculate fork choice", "err", err)
 		return
 	}
-	iterationInterval := time.NewTicker(30 * time.Millisecond)
+	iterationInterval := time.NewTicker(10000 * time.Millisecond)
```
Comment (Giulio2002) [inline cl/sentinel/gossip.go:195]: why?

File: cl/merkle_tree/merkle_tree.go [hunk 5]
```diff
@@ -54,6 +56,9 @@ func (m *MerkleTree) Initialize(leavesCount, maxTreeCacheDepth int, computeLeaf
 
 // MarkLeafAsDirty resets the leaf at the given index, so that it will be recomputed on the next call to ComputeRoot.
 func (m *MerkleTree) MarkLeafAsDirty(idx int) {
+	m.mu.Lock()
```
Comment (Giulio2002) [inline cl/merkle_tree/merkle_tree.go:59]: plz test performance impact on `ProcessEpoch`

File: cl/phase1/network/services/attestation_service_test.go [hunk 6]
```diff
@@ -81,7 +81,7 @@ func (t *attestationTestSuite) SetupTest() {
 	netConfig := &clparams.NetworkConfig{}
 	emitters := beaconevents.NewEventEmitter()
 	computeSigningRoot = func(obj ssz.HashableSSZ, domain []byte) ([32]byte, error) { return [32]byte{}, nil }
-	batchCheckInterval = 1 * time.Millisecond
+	batchCheckInterval.Store(1)
```
Comment (Giulio2002) [inline cl/phase1/network/services/attestation_service_test.go:84]: why?

**Review by Giulio2002 — CHANGES_REQUESTED** [review]

**Comment by Giulio2002** [comment]
Btw - concurrent read is not a "race" - i think you are atomicizing way too much

File: cl/beacon/handler/config.go [hunk 7]
```diff
@@ -41,7 +42,11 @@ func (a *ApiHandler) getDepositContract(w http.ResponseWriter, r *http.Request)
 func (a *ApiHandler) getForkSchedule(w http.ResponseWriter, r *http.Request) (*beaconhttp.BeaconResponse, error) {
 	response := []cltypes.Fork{}
 	// create first response (unordered and incomplete)
-	for currentVersion, entry := range a.beaconChainCfg.ForkVersionSchedule {
+	forkVersionSchedule := map[libcommon.Bytes4]clparams.VersionScheduleEntry{}
+	if a.beaconChainCfg.ForkVersionSchedule.Load() != nil {
+		forkVersionSchedule = *a.beaconChainCfg.ForkVersionSchedule.Load()
+	}
+	for currentVersion, entry := range forkVersionSchedule {
```
Comment (Giulio2002) [inline cl/beacon/handler/config.go:49]: this is called once at startup... how does this cause a race?

File: cl/clparams/config.go [hunk 8]
```diff
@@ -485,7 +486,7 @@ type BeaconChainConfig struct {
 	ElectraForkVersion   ConfigForkVersion `yaml:"ELECTRA_FORK_VERSION" spec:"true" json:"ELECTRA_FORK_VERSION"`        // ElectraForkVersion is used to represent the fork version for Electra.
 	ElectraForkEpoch     uint64            `yaml:"ELECTRA_FORK_EPOCH" spec:"true" json:"ELECTRA_FORK_EPOCH,string"`     // ElectraForkEpoch is used to represent the assigned fork epoch for Electra.
 
-	ForkVersionSchedule map[libcommon.Bytes4]VersionScheduleEntry `json:"-"` // Schedule of fork epochs by version.
+	ForkVersionSchedule atomic.Pointer[map[libcommon.Bytes4]VersionScheduleEntry] `json:"-"` // Schedule of fork epochs by version.
```
Comment (Giulio2002) [inline cl/clparams/config.go:489]: this is updated once at startup... how does this cause a race?

File: cl/clparams/config.go [hunk 9]
```diff
@@ -586,17 +587,19 @@ func (b *BeaconChainConfig) GetCurrentStateVersion(epoch uint64) StateVersion {
 
 // InitializeForkSchedule initializes the schedules forks baked into the config.
 func (b *BeaconChainConfig) InitializeForkSchedule() {
-	b.ForkVersionSchedule = configForkSchedule(b)
+	m := configForkSchedule(b)
+	b.ForkVersionSchedule.Store(&m)
 }
 
 func configForkSchedule(b *BeaconChainConfig) map[libcommon.Bytes4]VersionScheduleEntry {
```
Comment (Giulio2002) [inline cl/clparams/config.go:594]: this is called once at startup... how does this cause a race?

File: cl/clparams/config.go [hunk 10]
```diff
@@ -586,17 +587,19 @@ func (b *BeaconChainConfig) GetCurrentStateVersion(epoch uint64) StateVersion {
 
 // InitializeForkSchedule initializes the schedules forks baked into the config.
 func (b *BeaconChainConfig) InitializeForkSchedule() {
```
Comment (Giulio2002) [inline cl/clparams/config.go:589]: this is called once at startup... how does this cause a race?

**Review by Giulio2002 — CHANGES_REQUESTED** [review]

File: cl/phase1/network/services/batch_signature_verification.go (hunk 2 above)
Comment (Giulio2002) [reply cl/phase1/network/services/batch_signature_verification.go:45]: yeah no keep this as normal - i am definetely not doing this just for a race warn

File: cl/phase1/network/services/batch_signature_verification.go (hunk 3 above)
Comment (Giulio2002) [reply cl/phase1/network/services/batch_signature_verification.go:85]: yeah no keep this as normal - i am definetely not doing this just for a race warn

**Comment by domiwei** [comment]
I suggest to move `BatchSignatureVerifier` out of services pkg and mock it.

**Review by Giulio2002 — APPROVED** [review]

**Pull request merged**